File: sdk/typescript/hecate_sdk.ts
// =============================================================================
// HECATE REST API Client SDK — TypeScript Edition
// Provides typed interfaces to query incidents, approvals, and reports.
// =============================================================================

export interface Incident {
  id: string;
  incident_code: string;
  title: string;
  severity: string;
  status: string;
  service_name: string;
  root_cause: string;
  confidence_score: number;
  detected_at: string;
  resolved_at?: string;
  recovery_time_seconds?: number;
}

export interface Approval {
  id: string;
  incident_id: string;
  incident_type: string;
  recommended_action: string;
  root_cause_service: string;
  risk_level: string;
  recommendation_score: number;
  status: string;
  requested_at: string;
}

export class HecateClient {
  private baseUrl: string;
  private tenantId: string;

  constructor(baseUrl: string = 'http://localhost:8000', tenantId: string = 'default') {
    this.baseUrl = baseUrl;
    this.tenantId = tenantId;
  }

  private get headers(): HeadersInit {
    return {
      'X-Tenant-ID': this.tenantId,
      'Content-Type': 'application/json',
    };
  }

  async getIncidents(): Promise<Incident[]> {
    const res = await fetch(`${this.baseUrl}/api/v1/incidents`, {
      headers: this.headers,
    });
    if (!res.ok) throw new Error(`HTTP error ${res.status}`);
    return res.json();
  }

  async getApprovals(): Promise<Approval[]> {
    const res = await fetch(`${this.baseUrl}/api/v1/approvals`, {
      headers: this.headers,
    });
    if (!res.ok) throw new Error(`HTTP error ${res.status}`);
    return res.json();
  }

  async resolveApproval(
    approvalId: string,
    action: string,
    operator: string = 'operator'
  ): Promise<any> {
    const res = await fetch(`${this.baseUrl}/api/v1/approvals/${approvalId}/resolve`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({ action, operator }),
    });
    if (!res.ok) throw new Error(`HTTP error ${res.status}`);
    return res.json();
  }

  async injectChaos(faultType: string, serviceName: string): Promise<any> {
    const res = await fetch(`${this.baseUrl}/api/v1/chaos/inject`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({ fault_type: faultType, service_name: serviceName }),
    });
    if (!res.ok) throw new Error(`HTTP error ${res.status}`);
    return res.json();
  }

  async recoverChaos(faultType: string, serviceName: string): Promise<any> {
    const res = await fetch(`${this.baseUrl}/api/v1/chaos/recover`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify({ fault_type: faultType, service_name: serviceName }),
    });
    if (!res.ok) throw new Error(`HTTP error ${res.status}`);
    return res.json();
  }

  async getWeeklyReport(): Promise<any> {
    const res = await fetch(`${this.baseUrl}/api/v1/reports/weekly`, {
      headers: this.headers,
    });
    if (!res.ok) throw new Error(`HTTP error ${res.status}`);
    return res.json();
  }
}

File: sdk/python/hecate_sdk.py
# =============================================================================
# HECATE REST API Client SDK — Python Edition
# Provides typed interfaces to query incidents, approvals, and reports.
# =============================================================================

import httpx
from typing import List, Dict, Any, Optional

class HecateClient:
    def __init__(self, base_url: str = "http://localhost:8000", tenant_id: str = "default"):
        self.base_url = base_url
        self.headers = {
            "X-Tenant-ID": tenant_id,
            "Content-Type": "application/json"
        }

    def get_incidents(self) -> List[Dict[str, Any]]:
        """Retrieve list of active and resolved incidents for the current tenant."""
        res = httpx.get(f"{self.base_url}/api/v1/incidents", headers=self.headers)
        res.raise_for_status()
        return res.json()

    def get_approvals(self) -> List[Dict[str, Any]]:
        """Retrieve list of pending and resolved approvals for the current tenant."""
        res = httpx.get(f"{self.base_url}/api/v1/approvals", headers=self.headers)
        res.raise_for_status()
        return res.json()

    def resolve_approval(self, approval_id: str, action: str, operator: str = "operator") -> Dict[str, Any]:
        """Approve or reject a pending self-healing remediation action."""
        payload = {"action": action, "operator": operator}
        res = httpx.post(f"{self.base_url}/api/v1/approvals/{approval_id}/resolve", json=payload, headers=self.headers)
        res.raise_for_status()
        return res.json()

    def inject_chaos(self, fault_type: str, service_name: str) -> Dict[str, Any]:
        """Inject an infrastructure fault (e.g. pod_crash) to test platform resilience."""
        payload = {"fault_type": fault_type, "service_name": service_name}
        res = httpx.post(f"{self.base_url}/api/v1/chaos/inject", json=payload, headers=self.headers)
        res.raise_for_status()
        return res.json()

    def recover_chaos(self, fault_type: str, service_name: str) -> Dict[str, Any]:
        """Recover an injected infrastructure fault."""
        payload = {"fault_type": fault_type, "service_name": service_name}
        res = httpx.post(f"{self.base_url}/api/v1/chaos/recover", json=payload, headers=self.headers)
        res.raise_for_status()
        return res.json()

    def get_weekly_report(self) -> Dict[str, Any]:
        """Retrieve SRE weekly report compiling MTTR, availability, and SLA indicators."""
        res = httpx.get(f"{self.base_url}/api/v1/reports/weekly", headers=self.headers)
        res.raise_for_status()
        return res.json()
